src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -3097,24 +3103,92 @@ void addObjectEquals(MethodSymbol thisEquals)
+                    {
+                        return property;
+                    }
+                }
+                return null;
+            }
+
+            PropertySymbol addEqualityContract()
+            {
+                var property = new SynthesizedRecordEqualityContractProperty(this, isOverride: getInheritedEqualityContract(this) is object);
+                // https://github.com/dotnet/roslyn/issues/44903: Check explicit member has expected signature.
+                if (!memberSignatures.ContainsKey(property))
+                {
+                    members.Add(property);
+                    members.Add(property.GetMethod);
+                }
+                return property;
+            }
+
+            MethodSymbol addThisEquals(PropertySymbol equalityContract, MethodSymbol? otherEqualsMethod)
+            {
+                var thisEquals = new SynthesizedRecordEquals(this, parameterType: this, isOverride: false, equalityContract, otherEqualsMethod, memberOffset: members.Count);
                 if (!memberSignatures.TryGetValue(thisEquals, out var existing))
                 {
                     members.Add(thisEquals);
                     return thisEquals;
                 }
                 return (MethodSymbol)existing;
             }
+
+            static void getOtherEquals(ArrayBuilder<MethodSymbol> otherEqualsMethods, PropertySymbol equalityContract)
+            {
+                while ((equalityContract = equalityContract.OverriddenProperty) is object)
+                {
+                    var member = equalityContract.ContainingType.GetMembers("Equals").FirstOrDefault(m =>
+                    {
+                        if (m is MethodSymbol method)
+                        {
+                            var parameters = method.Parameters;
+                            if (parameters.Length == 1 && parameters[0].Type.Equals(m.ContainingType, TypeCompareKind.AllIgnoreOptions))
+                            {
+                                return true;
+                            }
+                        }
+                        return false;
+                    });
+                    // https://github.com/dotnet/roslyn/issues/44903: Check explicit member has expected signature.
+                    if (member is MethodSymbol method)
+                    {
+                        otherEqualsMethods.Add(method);
+                    }
+                }
+            }
+
+            void addOtherEquals(ArrayBuilder<MethodSymbol> otherEqualsMethods, PropertySymbol equalityContract, MethodSymbol thisEquals)
+            {
+                foreach (var otherEqualsMethod in otherEqualsMethods)
+                {
+                    var method = new SynthesizedRecordEquals(
+                        this,
+                        parameterType: otherEqualsMethod.Parameters[0].Type,
```

Comment:
Consider the following case

```
record A<T> { } 
record B : A<int> { }
```

Will `otherEqualsMethod.Parameters[0].Type` be the closed or open generic here?

Reply:
Great test case, thanks. It looks like it's handled correctly but I'll add a test. If there are no additional changes to this PR, perhaps I could add it in a subsequent PR.